### PR TITLE
Keep RemoteCluster resources up-to-date in the backend

### LINKED ISSUE #3869: "tctl get rc" doesn't show correct last_heartbeat for offline clusters
### Description

**What happened**:

Running `tctl get rc` returns the zero time for offline clusters:
```
$ tctl get rc
kind: remote_cluster
metadata:
  name: b
status:
  connection: offline
  last_heartbeat: "0001-01-01T00:00:00Z"
version: v3
```

**What you expected to happen**:

`last_heartbeat` should show the last time that remote cluster was connected.

**How to reproduce it (as minimally and precisely as possible)**:

- start cluster A
- start cluster B
- create a trusted cluster resource in cluster B for cluster A
- run `tctl get rc` on cluster A
- stop cluster B
- run `tctl get rc` on cluster A

### PR DESCRIPTION
**Pull request by awly**
Prior to this PR `RemoteCluster` resource data was dynamically generated from `TunnelConnection` resources.

Keep using `TunnelConnections`, but record any changes to `RemoteCluster` in the backend too. This lets us preserve the `LastHeartbeat` field after `TunnelConnections` get deleted (as they do when a reverse tunnel disconnects).

Also, vendor a few testing dependencies for new tests following https://github.com/gravitational/teleport/blob/master/rfd/0001-testing-guidelines.md

Fixes https://github.com/gravitational/teleport/issues/3869

### CODE REVIEW

**Comment by webvictim** [comment]
Do you think this would have any bearing on https://github.com/gravitational/teleport/issues/3642?

**Comment by awly** [comment]
@webvictim I don't believe it will

**Comment by awly** [comment]
retest this please

File: lib/auth/clt.go [hunk 1]
```diff
@@ -1183,6 +1183,19 @@ func (c *Client) CreateRemoteCluster(rc services.RemoteCluster) error {
 	return trace.Wrap(err)
 }
 
+// UpdateRemoteCluster upates remote cluster resource
```
Comment (webvictim) [inline lib/auth/clt.go:1186]:
```suggestion
// UpdateRemoteCluster updates remote cluster resource
```

**Review by webvictim — COMMENTED** [review]
LGTM, I'll leave it for others

**Comment by webvictim** [comment]
retest this please

File: lib/auth/trustedcluster.go [hunk 2]
```diff
@@ -367,14 +368,35 @@ func (a *AuthServer) updateRemoteClusterStatus(remoteCluster services.RemoteClus
...
-		remoteCluster.SetConnectionStatus(tunnelStatus)
+	if err != nil {
+		if remoteCluster.GetConnectionStatus() != teleport.RemoteClusterStatusOffline {
+			remoteCluster.SetConnectionStatus(teleport.RemoteClusterStatusOffline)
+			if err := a.UpdateRemoteCluster(ctx, remoteCluster); err != nil {
+				return trace.Wrap(err)
+			}
+		}
+		return nil
+	}
```
Comment (fspmarshall) [inline lib/auth/trustedcluster.go:380]: Nit: might be more future-proof to not categorically suppress errors from `services.LatestTunnelConnection`.  Maybe either only suppress if `trace.IsNotFound(err)`, or move this logic to before the call the `services.LatestTunnelConnection` and scope it to `if len(connections) == 0` instead.

**Review by fspmarshall — APPROVED** [review]
LGTM, but note that we shouldn't put this into a release until #4183 is resolved since this PR has the potential to incorrectly mark `RemoteCluster` resources as offline until that is fixed.

File: lib/auth/apiserver.go [hunk 3]
```diff
@@ -128,6 +128,7 @@ func NewAPIServer(config *APIConfig) http.Handler {
 	srv.POST("/:version/server/credentials", srv.withAuth(srv.generateServerKeys))
 
 	srv.POST("/:version/remoteclusters", srv.withAuth(srv.createRemoteCluster))
+	srv.PUT("/:version/remoteclusters", srv.withAuth(srv.updateRemoteCluster))
```
Comment (fspmarshall) [inline lib/auth/apiserver.go:131]: Just realized; I don't actually see anywhere that this is called remotely.  Isn't `UpdateRemoteCluster` only ever called by the `AuthServer`?  Why is this being exposed as a remote API?

Comment (awly) [reply lib/auth/apiserver.go]: Ahh, I totally cargo-culted this.
You're right, we don't need this publicly available.
Removed.

File: lib/auth/trustedcluster.go (hunk 2 above)
Comment (awly) [reply lib/auth/trustedcluster.go]: Makes sense, added a `trace.IsNotFound` check

**Review by awly — COMMENTED** [review]
> LGTM, but note that we shouldn't put this into a release until #4183 is resolved since this PR has the potential to incorrectly mark RemoteCluster resources as offline until that is fixed.

@fspmarshall this would happen when reading `TunnelConnections`, right?
If so, it's already a potential problem, since user-facing status is inferred from tunnel connections at runtime. The only difference is that it would get recorded in the backend too.

**Comment by fspmarshall** [comment]
> status is inferred from tunnel connections at runtime. The only difference is that it would get recorded in the backend too.

@awly Fair point; I wasn't considering the fact that the update was being run on *every* load anyway.  Carry on.

**Comment by awly** [comment]
retest this please

**Comment by awly** [comment]
retest this please

**Pull request merged**